08-generic_cache/04_improvements/cache.go

```diff
@@ -58,63 +57,53 @@ func (c *Cache[K, V]) Read(key K) (V, error) {
 // Upsert overrides the value for a given key.
 func (c *Cache[K, V]) Upsert(key K, value V) {
 	// Lock the writing on the map
-	c.dataMutex.Lock()
-	defer c.dataMutex.Unlock()
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
 	// If the data is already present
 	_, alreadyPresent := c.data[key]
 	switch {
 	case alreadyPresent:
-		// Mark key as the last inserted.
+		// Discard any previous reference.
 		c.deleteKeyValue(key)
-		c.addKeyValue(key, value)
 	case len(c.data) == c.maxSize:
 		// There is no room left in our map.
 		// We need to remove an item to write a new one.
-		c.deleteOldest()
-		c.addKeyValue(key, value)
-	default:
-		// Inserting a new item - no special action to perform.
-		c.addKeyValue(key, value)
+		c.deleteKeyValue(c.chronologicalKeys[0])
 	}
+
+	// Finally, insert the item.
+	c.addKeyValue(key, value)
 }
 
 // Delete removes the entry for the given key.
 func (c *Cache[K, V]) Delete(key K) {
 	// Lock the deletion on the map
-	c.dataMutex.Lock()
-	defer c.dataMutex.Unlock()
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
 	c.deleteKeyValue(key)
 }
 
+// addKeyValue inserts a key and its value into the cache.
 func (c *Cache[K, V]) addKeyValue(key K, value V) {
 	c.data[key] = entryWithTimeout[V]{
 		value:   value,
 		expires: time.Now().Add(c.ttl),
 	}
-	c.oldestKeys = append(c.oldestKeys, key)
+	c.chronologicalKeys = append(c.chronologicalKeys, key)
 }
 
+// deleteKeyValue removes a key and its associated value from the cache.
 func (c *Cache[K, V]) deleteKeyValue(key K) {
 	// Find the key in the list.
-	for index, k := range c.oldestKeys {
+	for i, k := range c.chronologicalKeys {
```

Comment:
You can use https://pkg.go.dev/slices#Index

Reply:
This (and Delete) require go 1.21 - it's one of the things we'll have to update after the official release, I suppose. We'll ask Doug what's best - this chapter'll be out either very soon before or very soon after the 1.21 release

Reply:
1.21 will release in [august](https://tip.golang.org/doc/go1.21) -- it's currently in rc2. So, I definitely would recommend targeting 1.21.